### ISSUE TITLE
Provide argument definition from documentation in hover on function argument

### ISSUE THREAD

**Issue by renkun-ken** [REditorSupport/languageserver#108]
The help documentation in hover message of function tokens seem to be very long. I'm wondering if it is possible to extract argument definition from the documentation and provide it in function argument completion and hover of e.g. `header=` in the following example:

```r
read.csv("test.csv", header = FALSE)
```

**Comment by randy3k** [comment]
You could do it, but it requires some anti-engineering.

```
library(purrr)
topic <- "parse"
arg <- "encoding"
rd <- utils:::.getHelpFile(utils::help((topic)))
argtxt <- rd %>% keep(~attr(., "Rd_tag") == "\\arguments") %>% `[[`(1) %>% 
	keep(~attr(., "Rd_tag") == "\\item") %>% keep(~.[[1]][[1]][[1]] == arg)
argtxt[[1]][[2]] %>% map_if(~attr(., "Rd_tag") == "TEXT", ~.[[1]])
```

Almost there....

**Comment by renkun-ken** [comment]
I tried 

```r
topic <- "read.csv"
arg <- "header"
```

and it seems that `.[[1]][[1]][[1]]` needs to be checked like

```r
keep(~attr(., "Rd_tag") == "\\item" && 
      length(.[[1]]) && length(.[[1]][[1]]) && length(.[[1]][[1]][[1]]) && .[[1]][[1]][[1]] == arg)
```

**Comment by randy3k** [comment]
ah....we need more testings

**Comment by randy3k** [comment]
and this kind of things cannot be done cleanly without `purrr`.

**Comment by renkun-ken** [comment]
Yes, I didn't know there's structured objects like this for documentation. Such deeply recursive structures do not look nice but help a lot for the functionality.

**Comment by randy3k** [comment]
I might be able to come up with something nicer later (but not before next week). Or do you want to play with it?

**Comment by renkun-ken** [comment]
We may cache well-defined structures that are transformed from this so that we can easily pick up information from it.

**Comment by renkun-ken** [comment]
I'm looking forward to your nicer idea for this. I'll spend some time looking at #107 and see what I can do about it.

**Comment by renkun-ken** [comment]
The following code transforms the nested structure into a simple named list of character vectors, each argument corresponds to an element.

```r
topic <- "read.csv"
rd <- utils:::.getHelpFile(utils::help((topic)))
args <- NULL
items <- list()
for (item in rd) {
    if (attr(item, "Rd_tag") == "\\arguments") {
        args <- item
        break
    }
}
if (length(args)) {
    items <- args[vapply(args, function(arg) attr(arg, "Rd_tag") == "\\item", logical(1L))]
}
argnames <- vapply(items, function(item) {
    argname <- item[[1]][[1]]
    switch(attr(argname, "Rd_tag"),
        TEXT = argname, "\\dots" = "...", "")
}, character(1L))
names(items) <- argnames
argtxts <- lapply(items, function(item) {
    paste0(unlist(item[[2]], use.names = FALSE), collapse = "")
})
```

**Comment by randy3k** [comment]
Ha, you couldn't stand playing with it!

**Comment by randy3k** [comment]
closed by #111 and #112
